### PR TITLE
feat(spans): Send outcomes for spans

### PR DESCRIPTION
**Pull request by phacops**
We want to keep track of span outcomes in order to be able to bill them.

Since Relay is responsible to produce spans to Snuba directly, when we keep a span we need to also send an `Accepted` outcome.

### CODE REVIEW

**Review by jjbayer — CHANGES_REQUESTED** [review]
From the moment we `process` a span until it's actually sent to kafka, there are still a few failure modes (`StoreService::produce_span` is not infallible). So I think we should log the outcome right after producing to kafka, i.e. here: https://github.com/getsentry/relay/blob/4bccbbb9326ad0b1f77e7414120b2eeff8a940f5/relay-server/src/actors/store.rs#L849

`StoreService` already has access to `outcome_aggregator`, so it should be easy to track the outcome from there.

File: relay-server/src/envelope.rs
Comment (jjbayer) [inline relay-server/src/envelope.rs:620]: For standalone spans, `indexed` will always be `false`, because it gets its value from https://github.com/getsentry/relay/blob/518148061bde526b7f6713b8e3c32c4660e4d53d/relay-server/src/utils/managed_envelope.rs#L299, which is only true if there was an event!

Can we live with that for now? I can try to make `indexed` depend on an item header instead of the event flag, then everything should fall into place.

File: relay-base-schema/src/data_category.rs
Comment (jjbayer) [inline relay-base-schema/src/data_category.rs]: As discussed: We need to release librelay with this new category and then bump it in sentry and snuba before we start producing outcomes with this ID.

**Review by jjbayer — COMMENTED** [review]
We have a bunch of integration tests for outcomes in `test_outcome.py`, especially around indexed vs. non-indexed. Would be good to add one for spans.

File: relay-base-schema/src/data_category.rs
Comment (phacops) [reply relay-base-schema/src/data_category.rs]: https://github.com/getsentry/relay/pull/2937

File: relay-server/src/envelope.rs
Comment (phacops) [reply relay-server/src/envelope.rs:620]: I tried something in that regard, let me know what you think (https://github.com/getsentry/relay/pull/2930/commits/872eb2e16e6afda0e15157e60d661a31739f5734, https://github.com/getsentry/relay/pull/2930/commits/5b2974396314e510d6f5149b80311d8043c528d1).

File: relay-server/src/actors/processor/span/processing.rs
Comment (jjbayer) [inline relay-server/src/actors/processor/span/processing.rs]: This should take the value of the old item, not of the event.

File: relay-server/src/actors/processor/span/processing.rs
Comment (jjbayer) [inline relay-server/src/actors/processor/span/processing.rs:137]: This raises a new question: Do we want outcomes for spans that were extracted from transactions?

This question is especially relevant for the `Accepted` outcome. When it becomes time to charge customers for standalone spans, won't we need to differentiate between spans that were submitted as part of transactions vs. spans that were submitted standalone? cc @indragiek

File: relay-server/src/actors/processor/span/processing.rs
Comment (jjbayer) [inline relay-server/src/actors/processor/span/processing.rs]:
```suggestion
        // If metrics extraction happened for the event, it also happened for its spans:
        item.set_metrics_extracted(state.event_metrics_extracted);
```

File: relay-server/src/utils/managed_envelope.rs
Comment (jjbayer) [inline relay-server/src/utils/managed_envelope.rs]:
```suggestion
```

File: relay-server/src/utils/managed_envelope.rs
Comment (jjbayer) [inline relay-server/src/utils/managed_envelope.rs]:
```suggestion
```

File: relay-server/src/utils/managed_envelope.rs
Comment (jjbayer) [inline relay-server/src/utils/managed_envelope.rs]:
```suggestion
                let use_indexed = if item.ty() == &ItemType::Span {
                    item.metrics_extracted()
                } else {
                    use_indexed
                };
                if let Some(category) = item.outcome_category(use_indexed) {
                    outcomes.push((outcome, category, quantity));
                };

                
```

File: relay-server/src/actors/processor/span/processing.rs
Comment (phacops) [reply relay-server/src/actors/processor/span/processing.rs:137]: There's no difference between charging for extracted or standalone spans.

Once the customer choose the new plan, we'll only look at span outcomes to understand how to bill them. The customer might use a newer SDK, sending standalone spans, or an older SDK, sending transactions and in both cases, we'd want to know the number of accepted spans to bill them and that's why we want to count spans extracted from the transactions as well.

Comment (jjbayer) [reply relay-server/src/actors/processor/span/processing.rs:137]: Alright!

File: relay-server/src/actors/store.rs
Comment (jjbayer) [inline relay-server/src/actors/store.rs]: Let's set these fields (`event_id` and `remote_addr`) to `None` so that the outcome aggregator can summarize outcomes. It already drops these fields under certain conditions (e.g. `Invalid` outcome), but spans are very high volume so we don't want to ever generate a separate outcome kafka message for every span.

```suggestion
            event_id: None,
            outcome: Outcome::Accepted,
            quantity: 1,
            remote_addr: None,
```

**Review by jjbayer — APPROVED** [review]
Approving to unblock. Before you merge, please

* Remove `event_id` and `remote_addr` from the `Accepted` outcome.
* Make sure librelay versions in sentry and snuba are up-to-date with the new `SpanIndexed` category.
* Add an integration test.
* After merging, deploy to Canary and observe how many extra Kafka messages are being sent.

File: relay-server/src/actors/store.rs
Comment (jjbayer) [reply relay-server/src/actors/store.rs]: Actually, could you also add `Accepted` to the `matches!` condition here, so that we don't mess this up in any future use cases of accepted? Thanks! https://github.com/getsentry/relay/blob/5cd6c1ed620f0ae6cf5823b98245c32582d45f60/relay-server/src/actors/outcome_aggregator.rs#L163-L172

**Pull request merged**
